pennylane/workflow/set_shots.py

```diff
@@ -16,22 +16,32 @@
 """
 from __future__ import annotations
 
-from collections.abc import Callable, Sequence
-from typing import TYPE_CHECKING, Optional
+from collections.abc import Callable
+from typing import TYPE_CHECKING, overload
 
 from .qnode import QNode
 
-# Sentinel value to distinguish decorator mode from direct calls
-_DECORATOR_MODE = object()
-
 if TYPE_CHECKING:
-    from pennylane.measurements import Shots
+    from pennylane.measurements import ShotsLike
+
+
+# Sentinel value to detect when shots parameter is not provided
+_SHOTS_NOT_PROVIDED = object()
+
+
+@overload
+def set_shots(qnode: QNode, shots: ShotsLike) -> QNode: ...
+
+
+@overload
+def set_shots(shots: ShotsLike) -> Callable[[QNode], QNode]: ...
+
+
+@overload
+def set_shots(*, shots: ShotsLike) -> Callable[[QNode], QNode]: ...
 
 
-def set_shots(
-    qnode: QNode | object = _DECORATOR_MODE,
-    shots: Optional[Shots | int | Sequence[int | tuple[int, int]]] = None,
-) -> QNode | Callable[[QNode], QNode]:
+def set_shots(*args, shots: ShotsLike = _SHOTS_NOT_PROVIDED):
```

Comment:
```suggestion
@overload
def set_shots(qnode: QNode, shots: ShotsLike) -> QNode: ...
@overload
def set_shots(shots: ShotsLike) -> Callable[[QNode], QNode]: ...
@overload
def set_shots(*, shots: ShotsLike) -> Callable[[QNode], QNode]: ...
def set_shots(*args, shots: ShotsLike = _SHOTS_NOT_PROVIDED):
```
I don't think you typically leave space between the overloads